libbeat/outputs/elasticsearch/json_read.go

```diff
@@ -0,0 +1,545 @@
+package elasticsearch
+
+import (
+	"errors"
+
+	"github.com/elastic/beats/libbeat/common/streambuf"
+)
+
+// SAX like json parser. But instead of relying on callbacks, state machine
+// returns raw item plus entity. On top of state machine additional helper methods
+// like expectDict, expectArray, nextFieldName and nextInt are available for
+// low-level parsing/stepping through a json document.
+//
+// Due to parser simply stepping through the input buffer, almost no additional
+// allocations are required.
+type jsonReader struct {
+	streambuf.Buffer
+
+	// parser state machine
+	states       []state // state stack for nested arrays/objects
+	currentState state
+
+	// preallocate stack memory for up to 32 nested arrays/objects
+	statesBuf [32]state
+}
+
+var (
+	errFailing             = errors.New("JSON parser failed")
+	errUnknownChar         = errors.New("unknown character")
+	errQuoteMissing        = errors.New("missing closing quote")
+	errExpectColon         = errors.New("expected ':' after map key")
+	errUnexpectedDictClose = errors.New("unexpected '}'")
+	errUnexpectedArrClose  = errors.New("unexpected ']'")
+	errExpectedDigit       = errors.New("expected a digit")
+	errExpectedObject      = errors.New("expected JSON object")
+	errExpectedArray       = errors.New("expected JSON array")
+	errExpectedFieldName   = errors.New("expected JSON object field name")
+	errExpectedInteger     = errors.New("expected integer value")
+	errExpectedNull        = errors.New("expected null value")
+	errExpectedFalse       = errors.New("expected false value")
+	errExpectedTrue        = errors.New("expected true value")
+	errExpectedArrayField  = errors.New("expected ']' or ','")
+)
+
+var (
+	nullSymbol  = []byte("null")
+	trueSymbol  = []byte("true")
+	falseSymbol = []byte("false")
+)
+
+type entity uint8
+
+const (
+	failEntity entity = iota
+	trueValue
+	falseValue
+	nullValue
+	dictStart
+	dictEnd
+	dictField
+	arrStart
+	arrEnd
+	stringEntity
+	mapKeyEntity
+	intEntity
+	doubleEntity
+)
+
+type state uint8
+
+const (
+	failedState state = iota
+	startState
+	arrState
+	arrStateNext
+	dictState
+	dictFieldState
+	dictFieldStateEnd
+)
+
+func (s state) String() string {
+	switch s {
+	case failedState:
```

Comment:
Consider declaring a `map[state]string` to hold the string representations rather than using a switch. It makes the code coverage numbers look better since it lowers the number of testable lines.